### ISSUE TITLE
Assertion error when building in non-strict mode

### ISSUE THREAD

**Issue by Cynical-Optimist** [buildstream-migration/bst-staging#316]
[See original issue on GitLab](https://gitlab.com/BuildStream/buildstream/-/issues/316)
In GitLab by [[Gitlab user @abderrahimk]](https://gitlab.com/abderrahimk) on Mar 24, 2018, 15:09

After I did a change early in the dependency chain and didn't want to wait for the build to complete, I interrupted it and disabled strict mode. I got the following error. The element in question is a compose
```
[--:--:--][????????][build:base/freedesktop-sdk-junction.bst:desktop-image.bst] START   Caching Artifact
[00:12:11][????????][build:base/freedesktop-sdk-junction.bst:desktop-image.bst] BUG     Build

    An unhandled exception occured:
    
    Traceback (most recent call last):
      File "/home/abderrahim/dev/buildstream/buildstream/_scheduler/job.py", line 265, in child_action
        result = self.action(element)
      File "/home/abderrahim/dev/buildstream/buildstream/_scheduler/buildqueue.py", line 34, in process
        element._assemble()
      File "/home/abderrahim/dev/buildstream/buildstream/element.py", line 1149, in _assemble
        self.__artifacts.commit(self, assembledir, self.__get_cache_keys_for_commit())
      File "/home/abderrahim/dev/buildstream/buildstream/_artifactcache/ostreecache.py", line 265, in commit
        refs = [buildref(element, key) for key in keys]
      File "/home/abderrahim/dev/buildstream/buildstream/_artifactcache/ostreecache.py", line 265, in <listcomp>
        refs = [buildref(element, key) for key in keys]
      File "/home/abderrahim/dev/buildstream/buildstream/_artifactcache/ostreecache.py", line 47, in buildref
        assert key is not None
    AssertionError
```

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @abderrahimk]](https://gitlab.com/abderrahimk) on Mar 24, 2018, 15:09

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @abderrahimk]](https://gitlab.com/abderrahimk) on Mar 24, 2018, 17:12

I noticed it in another element (which is not a compose), and building with `--strict` works around the issue.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Mar 24, 2018, 17:29

Im on my handphone right now, there is another similar issue reported by Michael which hits the same assertion.

Are you able to consistently reproduce this ?

Could you send us at least the project in the state which causes this ?

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @abderrahimk]](https://gitlab.com/abderrahimk) on Mar 24, 2018, 17:36

I've already done a strict build now, I will try to reproduce it later.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @abderrahimk]](https://gitlab.com/abderrahimk) on Apr 4, 2018, 15:10

I managed to reproduce it. I did a `git add -f .`, I'm now pushing. Should appear [here](https://gitlab.gnome.org/abderrahim/gnome-build-meta/branches) in a few minutes. I hope it is enough to reproduce the error. I'm afraid it also depends on the contents of `~/.cache/buildstream/artifacts/`.

The command I ran is `bst build core-deps/samba.bst`.

Something weird I noticed as well: in the pipeline listing before the build start, I see this
```
buildable ???????? core-deps/samba.bst Workspaces: ~/dev/gnome-build-meta/samba, ~/dev/gnome-build-meta/samba
```
(the workspace is printed twice) even though .bst/workspaces.yml lists it only once.

I can try to debug it if you can tell me where I should look.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Apr 4, 2018, 15:26

The fact that it's showing two workspaces there seems to be a regression from making workspaces source based to element based.

It seems that your `samba.bst` has a tarball and a patch, and the frontend is listing two workspaces mistakenly, this would be unrelated.

Is it possible to run this with `bst --log-file build.log build core-deps/samba.bst`, and attach the log file as an attachment here ?

That might be able to give us a clue as to what was cached or not at the time of the build.

In the meanwhile, I'm trying a build of this...

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @abderrahimk]](https://gitlab.com/abderrahimk) on Apr 4, 2018, 16:42

Here is the exact state of my workdir https://gitlab.gnome.org/abderrahim/gnome-build-meta/tree/buildstream-bug (you can ignore freedesktop as I closed the workspace). If there is anything else I can try, please tell me.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Apr 5, 2018, 05:39

I reproduced with your branch !

Looks like we're on to something here...

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Apr 5, 2018, 07:01

mentioned in commit d6a47a042183ca4e906f88022a732f9562d04a52

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Apr 5, 2018, 07:01

mentioned in commit e52f1a98281fa3344f694f1c4325900c58aa9e1b

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Apr 5, 2018, 07:04

mentioned in merge request !358

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Apr 5, 2018, 07:16

mentioned in commit 9012a5ed2f94704a99b3a2ac8f673160fca86668

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Apr 5, 2018, 07:16

mentioned in commit daffe49827e1d462a6aa230afcd188645198dfea

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Apr 5, 2018, 07:34

closed via merge request !358

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @abderrahimk]](https://gitlab.com/abderrahimk) on Apr 27, 2018, 10:47

I had this error yesterday with 1.1.3, and as in my original report it wasn't in a workspaced element. (to be sure, I closed all workspaces and deleted .bst).

Ironically, it succeeded when I retried with --log-file to provide the log. The actual difference wasn't --log-file though. I was offline in the second attempt, so it didn't connect to the artifact cache.

I'm attaching the result of both [failed-build.txt](/uploads/768ca5cba8429c89f0d4496b988fd483/failed-build.txt) and (truncated) [succeeded-build.txt](/uploads/77684e35f2f2836025ca818bd604fdeb/succeeded-build.txt), copied from the terminal.

My project state is [here](https://gitlab.gnome.org/GNOME/gnome-build-meta/tree/abderrahim/buildstream-bug).

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @abderrahimk]](https://gitlab.com/abderrahimk) on Apr 27, 2018, 10:47

reopened

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @toscalix]](https://gitlab.com/toscalix) on May 18, 2018, 15:24

assigned to [[Gitlab user @abderrahimk]](https://gitlab.com/abderrahimk)

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @toscalix]](https://gitlab.com/toscalix) on May 18, 2018, 15:26

assigned to [[Gitlab user @toscalix]](https://gitlab.com/toscalix) and unassigned [[Gitlab user @abderrahimk]](https://gitlab.com/abderrahimk)

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @toscalix]](https://gitlab.com/toscalix) on Aug 20, 2018, 09:28

assigned to [[Gitlab user @valentindavid]](https://gitlab.com/valentindavid)

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @toscalix]](https://gitlab.com/toscalix) on Aug 23, 2018, 11:13

unassigned [[Gitlab user @toscalix]](https://gitlab.com/toscalix)

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @valentindavid]](https://gitlab.com/valentindavid) on Aug 23, 2018, 11:43

Based on the logs:

- python3 did not have a strong key.
- python3 was cached with weak key.
- meson failed to be committed because I suppose it could not calculate the strong key which depends on python3's strong key.
 
For python3 to not have a strong key but have a weak key and being cached is maybe because python3 depends on an element with workspace that has been opened but not built. It is not build because of `--no-strict`.

It seems to me that it is possible to not have strong key in this case.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @valentindavid]](https://gitlab.com/valentindavid) on Aug 23, 2018, 11:47

I suspect that could trigger the issue:

 - Create element `a.bst`
 - Create `b.bst` that build depends on `a.bst`
 - Create `c.bst` that build depends on `b.bst`
 - Build `b.bst`.
 - Open workspace for `a.bst`.
 - Build `c.bst` with `--no-strict`.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @valentindavid]](https://gitlab.com/valentindavid) on Aug 24, 2018, 13:20

That did not work. Now this is what I think happens:

You must have an artifact server. This is important because it changes the behavior of keys.

Have `a.bst`, `b.bst` that build depends on `a.bst` and `c.bst` that build depends on `b.bst`.

`b.bst` and `c.bst` should not use a plugin with `BST_STRICT_REBUILD`. So no `compose`.

Build `b.bst`.

Modify `a.bst` so the key changes.

Build `c.bst`.

At this point BuildStream will want to pull `a.bst` and `b.bst`. Which I think is a bug by itself, we already have `b.bst` with the right weak key. Note that the pipeline summary will show `????` for all three elements.

It seems that even if `a.bst` and `b.bst` will be pulled from the cache. `c.bst` will start being built before it is finished.

So there, we have 2 possibility. If the pull from the server `b.bst`, then we will have a strong key for `c.bst` for the pulled artifact. We have however built `b.bst` with the older artifact. But we will commit it with the computed strong key depending on the new `b.bst`.

Second possibility. The server is slow. We finish building `c.bst`. Then we try to commit without a strong key (which by the way, we should have because we have the strong keys of all artifacts we used). Then we break an assertion that we try to commit with an invalid key.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @valentindavid]](https://gitlab.com/valentindavid) on Aug 24, 2018, 13:21

[[Gitlab user @juergbi]](https://gitlab.com/juergbi) Any opinion on that?

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Aug 28, 2018, 13:01

> At this point BuildStream will want to pull `a.bst` and `b.bst`. Which I think is a bug by itself, we already have `b.bst` with the right weak key.

That's by design. It will attempt to pull an artifact from a strict build to keep possible non-strict build issues to a minimum.

> It seems that even if `a.bst` and `b.bst` will be pulled from the cache. `c.bst` will start being built before it is finished.

At the end of `Element._update_state()`, the strong cache key is not set if pull is pending (and we're running in non-strict mode). This should mean that reverse dependencies are not `buildable` until the pull (attempt) is complete and the strong cache key is finally set.

Based on the above I don't currently see how this can happen. Maybe this issue was the same as #383, which was fixed a while ago.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Aug 28, 2018, 17:32

We expect this to be a duplicate of #383 and thus already fixed for a while now. If anyone can still reproduce this with a recent version of BuildStream, please reopen.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Aug 28, 2018, 17:32

closed
